keps/sig-node/20190528-node-maintenance-controller.md

```diff
@@ -0,0 +1,190 @@
+of the targeted release**.
+
+For enhancements that make changes to code or processes/procedures in core Kubernetes i.e., [kubernetes/kubernetes], we require the following Release Signoff checklist to be completed.
+
+Check these off as they are completed for the Release Team to track. These checklist items _must_ be updated for the enhancement to be released.
+
+- [ ] kubernetes/enhancements issue in release milestone, which links to KEP (this should be a link to the KEP location in kubernetes/enhancements, not the initial KEP PR)
+- [ ] KEP approvers have set the KEP status to `implementable`
+- [ ] Design details are appropriately documented
+- [ ] Test plan is in place, giving consideration to SIG Architecture and SIG Testing input
+- [ ] Graduation criteria is in place
+- [ ] "Implementation History" section is up-to-date for milestone
+- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
+- [ ] Supporting documentation e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+**Note:** Any PRs to move a KEP to `implementable` or significant changes once it is marked `implementable` should be approved by each of the KEP approvers. If any of those approvers is no longer appropriate than changes to that list should be approved by the remaining approvers and/or the owning SIG (or SIG-arch for cross cutting KEPs).
+
+**Note:** This checklist is iterative and should be reviewed and updated every time this enhancement is being considered for a milestone.
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://github.com/kubernetes/enhancements/issues
+[kubernetes/kubernetes]: https://github.com/kubernetes/kubernetes
+[kubernetes/website]: https://github.com/kubernetes/website
+
+## Summary
+
+Node maintenance controller that performs node drain - server side using custom resources calls to the API in order to perform the drain (cordon + evictions)
+
+## Motivation
+
+Move kubectl drain into the server to enable execution of node drain using custom resources calls to the API in order to perform the drain.
+This would also reduce the proliferation of drain behaviours in other projects that re-implement the drain logic to accustom their needs.
+
+(https://github.com/kubernetes/kubernetes/issues/25625)
+
+### Goals
+
+- A new controller that implements the drain logic that is currently [implemented](https://github.com/kubernetes/kubernetes/blob/v1.14.0/pkg/kubectl/cmd/drain/drain.go#L307) in [kubectl](https://kubernetes.io/docs/reference/generated/kubectl/kubectl-commands#drain).
+- The controller watches and acts upon maintenance CRs to put/remove a node on/from maintenance.
```

Comment:
The CRs seems a command to k8s cluster, is that only for maintenance?

Reply:
The CR can be interpreted as a different way to trigger drain using k8s API instead of CLI, opening new options including adding on top of the CR additional info such as: status,reason, custom taints/labels, etc.

Reply:
oh, I mean introducing a common CR, named as `Command`, for such kind of feature :)

Reply:
ah, yes. Here is an example for this common cr:
https://github.com/kubevirt/node-maintenance-operator/blob/master/deploy/crds/nodemaintenance_crd.yaml